File: highspeed_report/highspeed_stock/report/simple_items_barcode/simple_items_barcode.js
frappe.query_reports["Simple Items Barcode"] = {
    "filters": [
        {
            "fieldname": "item_code",
            "label": __("Item Code"),
            "fieldtype": "Link",
            "options": "Item"
        },
        {
            "fieldname": "item_group",
            "label": __("Item Group"),
            "fieldtype": "Link",
            "options": "Item Group"
        },
        {
            "fieldname": "barcode",
            "label": __("Barcode"),
            "fieldtype": "Data"
        },
        {
            "fieldname": "price_list",
            "label": __("Price List"),
            "fieldtype": "Link",
            "options": "Price List",
            "default": frappe.defaults.get_default("selling_price_list"),
            "get_query": function() {
                return {
                    filters: {
                        "selling": 1,
                        "enabled": 1
                    }
                };
            }
        },
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company")
        }
    ],

    "formatter": function(value, row, column, data, default_formatter) {
        if (!data) return default_formatter(value, row, column, data);
        
        if (value === null || value === undefined || value === "null") {
            value = "";
        }
        
        if (column.fieldname === "barcode" && value) {
            return `<span style="font-family: 'Courier New', monospace; font-weight: bold; font-size: 1.1em;">${value}</span>`;
        }
        
        if (column.fieldname === "selling_price") {
            let price = parseFloat(value) || 0;
            let currency = data.currency || frappe.defaults.get_default("currency");
            let formatted = format_currency(price, currency);
            
            let color = price > 0 ? "#2ecc71" : "#e74c3c";
            return `<span style="font-weight: bold; color: ${color};">${formatted}</span>`;
        }
        
        if (column.fieldname === "barcode_type" && value) {
            let typeColors = {
                "EAN": "#3498db",
                "UPC": "#e74c3c",
                "CODE128": "#f39c12",
                "CODE39": "#9b59b6",
                "QR": "#1abc9c"
            };
            let color = typeColors[value] || "#7f8c8d";
            return `<span style="color: ${color}; font-weight: 600;">${value}</span>`;
        }
        
        if (column.fieldname === "print_label") {
            return `<div style="text-align: center;">
                    <button class="btn btn-xs btn-primary print-single-label" 
                        data-item-code="${data.item_code}" 
                        data-item-name="${data.item_name}" 
                        data-barcode="${data.barcode}" 
                        data-price="${data.selling_price}">
                        <i class="fa fa-print"></i> طباعة
                    </button>
                    </div>`;
        }
        
        if (column.fieldname === "print_barcode") {
            return `<div style="text-align: center;">
                    <button class="btn btn-xs btn-warning print-barcode-only" 
                        data-barcode="${data.barcode}">
                        <i class="fa fa-barcode"></i> شريط باركود
                    </button>
                    </div>`;
        }
        
        return default_formatter(value, row, column, data);
    },

    "onload": function(report) {
        report.page.add_inner_button(__("Export Barcodes"), function() {
            exportBarcodes(report);
        });
        
        frappe.ui.keys.add_shortcut({
            shortcut: 'ctrl+e',
            action: () => exportBarcodes(report),
            page: report.page,
            description: __('Export Barcodes')
        });
    },

    "after_datatable_render": function(datatable_obj, report) {
        if (!datatable_obj) return;
        
        window.current_report = report;
        
        if (report && report.data) {
            report.data.forEach((row, idx) => {
                if (!row.selling_price || parseFloat(row.selling_price) === 0) {
                    $(`.dt-row-${idx}`).css('background-color', 'rgba(231, 76, 60, 0.1)');
                }
            });
        }
        
        $(document).off('click', '.print-single-label').on('click', '.print-single-label', function(e) {
            e.preventDefault();
            e.stopPropagation();
            
            let $btn = $(this);
            let itemData = {
                item_code: $btn.attr('data-item-code'),
                item_name: $btn.attr('data-item-name'),
                barcode: $btn.attr('data-barcode'),
                price: parseFloat($btn.attr('data-price')) || 0
            };
            
            printSingleLabel(itemData);
        });
        
        $(document).off('click', '.print-barcode-only').on('click', '.print-barcode-only', function(e) {
            e.preventDefault();
            e.stopPropagation();
            
            let $btn = $(this);
            let itemData = {
                barcode: $btn.attr('data-barcode')
            };
            
            printBarcodeOnly(itemData);
        });
        
        $(document).off('click', '.dt-cell--col-5').on('click', '.dt-cell--col-5', function() {
            let barcode = $(this).text().trim();
            if (barcode) {
                showBarcodeDialog(barcode);
            }
        });
    }
};

function printSingleLabel(item) {
    frappe.show_alert({message: __("جاري طباعة الملصق..."), indicator: 'blue'});
    
    let company = frappe.defaults.get_user_default("Company") || "";
    
    let labelHTML = `
        <div class="label">
            <div class="content-wrapper">
                <div class="company-name">${company}</div>
                <div class="item-name">${item.item_name || ''}</div>
                <div class="barcode-container">
                    <svg class="barcode" id="barcode-svg"></svg>
                    <div class="barcode-text">${item.barcode || ''}</div>
                </div>
                <div class="price">${format_currency(item.price || 0, frappe.defaults.get_default("currency"))}</div>
            </div>
        </div>
    `;
    
    let html = `
        <!DOCTYPE html>
        <html dir="rtl">
        <head>
            <meta charset="UTF-8">
            <title>طباعة ملصق - ${item.item_name || ''}</title>
            <script src="https://cdn.jsdelivr.net/npm/jsbarcode@3.11.0/dist/JsBarcode.all.min.js"></script>
            <style>
                * {
                    margin: 0;
                    padding: 0;
                    box-sizing: border-box;
                }
                
                @page { 
                    size: 50mm 25mm;
                    margin: 0;
                }
                
                html, body {
                    width: 50mm;
                    height: 25mm;
                    margin: 0;
                    padding: 0;
                    overflow: hidden;
                }
                
                body {
                    font-family: 'Arial', 'Segoe UI', Tahoma, sans-serif;
                    direction: rtl;
                }
                
                .label {
                    width: 50mm;
                    height: 25mm;
                    position: relative;
                    display: flex;
                    align-items: center;
                    justify-content: center;
                    padding: 1mm;
                    box-sizing: border-box;
                }
                
                .content-wrapper {
                    width: 100%;
                    height: 100%;
                    display: flex;
                    flex-direction: column;
                    justify-content: space-around;
                    text-align: center;
                    align-items: center;
                }
                
                .company-name {
                    font-size: 6pt;
                    font-weight: bold;
                    color: #000;
                    line-height: 1;
                    margin: 0;
                }
                
                .item-name {
                    font-size: 6.5pt;
                    font-weight: bold;
                    color: #000;
                    line-height: 1;
                    max-height: 4mm;
                    overflow: hidden;
                    text-overflow: ellipsis;
                    display: -webkit-box;
                    -webkit-line-clamp: 2;
                    -webkit-box-orient: vertical;
                    margin: 0;
                }
                
                .barcode-container {
                    display: flex;
                    flex-direction: column;
                    justify-content: center;
                    align-items: center;
                    margin: 0;
                }
                
                .barcode {
                    max-width: 46mm;
                    max-height: 8mm;
                }
                
                .barcode-text {
                    font-size: 7pt;
                    font-family: 'Courier New', monospace;
                    margin: 0;
                    line-height: 1;
                    font-weight: bold;
                }
                
                .price {
                    font-size: 7pt;
                    font-weight: bold;
                    color: #000;
                    line-height: 1;
                    margin: 0;
                }
                
                @media print {
                    @page {
                        size: 50mm 25mm;
                        margin: 0;
                    }
                    
                    html, body {
                        width: 50mm !important;
                        height: 25mm !important;
                        padding: 0 !important;
                        margin: 0 !important;
                    }
                    
                    .label {
                        width: 50mm !important;
                        height: 25mm !important;
                        page-break-inside: avoid;
                        page-break-after: avoid;
                        page-break-before: avoid;
                    }
                }
            </style>
        </head>
        <body>
            ${labelHTML}
            <script>
                window.onload = function() {
                    try {
                        if ("${item.barcode}") {
                            JsBarcode("#barcode-svg", "${item.barcode}", {
                                format: "CODE128",
                                width: 1,
                                height: 30,
                                displayValue: false,
                                margin: 0
                            });
                        }
                    } catch(e) {
                        console.error('Error generating barcode:', e);
                    }
                    
                    setTimeout(function() {
                        window.print();
                        setTimeout(function() {
                            window.close();
                        }, 1000);
                    }, 100);
                };
            </script>
        </body>
        </html>
    `;
    
    try {
        let printWindow = window.open('', '_blank', 'width=400,height=300,resizable=yes,scrollbars=yes');
        if (printWindow) {
            printWindow.document.write(html);
            printWindow.document.close();
        } else {
            frappe.msgprint(__("يرجى السماح بالنوافذ المنبثقة لطباعة الملصق"));
        }
    } catch(e) {
        frappe.msgprint(__("حدث خطأ أثناء فتح نافذة الطباعة: ") + e.message);
    }
}

function printBarcodeOnly(item) {
    frappe.show_alert({message: __("جاري طباعة الباركود فقط..."), indicator: 'blue'});
    
    let labelHTML = `
        <div class="label">
            <div class="content-wrapper">
                <div class="barcode-container">
                    <svg class="barcode" id="barcode-svg"></svg>
                    <div class="barcode-text">${item.barcode || ''}</div>
                </div>
            </div>
        </div>
    `;
    
    let html = `
        <!DOCTYPE html>
        <html dir="rtl">
        <head>
            <meta charset="UTF-8">
            <title>طباعة باركود - ${item.barcode || ''}</title>
            <script src="https://cdn.jsdelivr.net/npm/jsbarcode@3.11.0/dist/JsBarcode.all.min.js"></script>
            <style>
                * {
                    margin: 0;
                    padding: 0;
                    box-sizing: border-box;
                }
                
                @page { 
                    size: 50mm 25mm;
                    margin: 0;
                }
                
                html, body {
                    width: 50mm;
                    height: 25mm;
                    margin: 0;
                    padding: 0;
                    overflow: hidden;
                }
                
                body {
                    font-family: 'Arial', 'Segoe UI', Tahoma, sans-serif;
                    direction: rtl;
                }
                
                .label {
                    width: 50mm;
                    height: 25mm;
                    position: relative;
                    display: flex;
                    align-items: center;
                    justify-content: center;
                    padding: 1mm;
                    box-sizing: border-box;
                }
                
                .content-wrapper {
                    width: 100%;
                    height: 100%;
                    display: flex;
                    flex-direction: column;
                    justify-content: center;
                    text-align: center;
                    align-items: center;
                }
                
                .barcode-container {
                    display: flex;
                    flex-direction: column;
                    justify-content: center;
                    align-items: center;
                    margin: 0;
                }
                
                .barcode {
                    max-width: 46mm;
                    max-height: 15mm;
                }
                
                .barcode-text {
                    font-size: 8pt;
                    font-family: 'Courier New', monospace;
                    margin: 2px 0 0 0;
                    line-height: 1;
                    font-weight: bold;
                }
                
                @media print {
                    @page {
                        size: 50mm 25mm;
                        margin: 0;
                    }
                    
                    html, body {
                        width: 50mm !important;
                        height: 25mm !important;
                        padding: 0 !important;
                        margin: 0 !important;
                    }
                    
                    .label {
                        width: 50mm !important;
                        height: 25mm !important;
                        page-break-inside: avoid;
                        page-break-after: avoid;
                        page-break-before: avoid;
                    }
                }
            </style>
        </head>
        <body>
            ${labelHTML}
            <script>
                window.onload = function() {
                    try {
                        if ("${item.barcode}") {
                            JsBarcode("#barcode-svg", "${item.barcode}", {
                                format: "CODE128",
                                width: 1.2,
                                height: 50,
                                displayValue: false,
                                margin: 0
                            });
                        }
                    } catch(e) {
                        console.error('Error generating barcode:', e);
                    }
                    
                    setTimeout(function() {
                        window.print();
                        setTimeout(function() {
                            window.close();
                        }, 1000);
                    }, 100);
                };
            </script>
        </body>
        </html>
    `;
    
    try {
        let printWindow = window.open('', '_blank', 'width=400,height=300,resizable=yes,scrollbars=yes');
        if (printWindow) {
            printWindow.document.write(html);
            printWindow.document.close();
        } else {
            frappe.msgprint(__("يرجى السماح بالنوافذ المنبثقة لطباعة الباركود"));
        }
    } catch(e) {
        frappe.msgprint(__("حدث خطأ أثناء فتح نافذة الطباعة: ") + e.message);
    }
}

function exportBarcodes(report) {
    if (!report.data || report.data.length === 0) {
        frappe.msgprint(__("No data to export"));
        return;
    }
    
    let export_data = [];
    export_data.push(["Item Code", "Item Name", "Item Group", "UOM", "Selling Price", "Barcode", "Barcode Type"]);
    
    report.data.forEach(row => {
        export_data.push([
            row.item_code || '',
            row.item_name || '',
            row.item_group || '',
            row.uom || '',
            row.selling_price || 0,
            row.barcode || '',
            row.barcode_type || ''
        ]);
    });
    
    frappe.tools.downloadify(export_data, null, 'Items_Barcode_Report');
}

function showBarcodeDialog(barcode) {
    let dialog = new frappe.ui.Dialog({
        title: __('Barcode: ') + barcode,
        fields: [
            {
                fieldtype: 'HTML',
                fieldname: 'barcode_display'
            }
        ]
    });
    
    let canvas = document.createElement('canvas');
    try {
        JsBarcode(canvas, barcode, {
            format: "CODE128",
            width: 2,
            height: 100,
            displayValue: true
        });
        
        dialog.fields_dict.barcode_display.$wrapper.html(`
            <div style="text-align: center; padding: 20px;">
                <img src="${canvas.toDataURL()}" alt="${barcode}" />
                <p style="margin-top: 10px; font-size: 14px;">
                    <button class="btn btn-primary btn-sm" onclick="navigator.clipboard.writeText('${barcode}')">
                        ${__('Copy Barcode')}
                    </button>
                </p>
            </div>
        `);
    } catch(e) {
        dialog.fields_dict.barcode_display.$wrapper.html(`
            <div style="text-align: center; padding: 20px; color: red;">
                ${__('Error generating barcode')}: ${e.message}
            </div>
        `);
    }
    
    dialog.show();
}

File: highspeed_report/highspeed_stock/report/simple_items_barcode/simple_items_barcode.py
# Copyright (c) 2025, Your Company and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, cstr

def execute(filters=None):
    """Main execution function for the Simple Items Barcode report"""
    if not filters:
        filters = {}
    
    # Get report data
    columns = get_columns()
    data = get_item_barcode_data(filters)
    
    # Add summary if needed
    summary = get_summary(data)
    
    return columns, data, None, None, summary

def get_columns():
    """Define the columns for the simple items barcode report"""
    return [
        {
            "fieldname": "item_code",
            "label": _("Item Code"),
            "fieldtype": "Link",
            "options": "Item",
            "width": 120
        },
        {
            "fieldname": "item_name",
            "label": _("Item Name"),
            "fieldtype": "Data",
            "width": 250
        },
        {
            "fieldname": "item_group",
            "label": _("Item Group"),
            "fieldtype": "Link",
            "options": "Item Group",
            "width": 120
        },
        {
            "fieldname": "uom",
            "label": _("UOM"),
            "fieldtype": "Link",
            "options": "UOM",
            "width": 80
        },
        {
            "fieldname": "selling_price",
            "label": _("Selling Price"),
            "fieldtype": "Currency",
            "width": 120,
            "options": "currency"
        },
        {
            "fieldname": "barcode",
            "label": _("Barcode"),
            "fieldtype": "Data",
            "width": 150
        },
        {
            "fieldname": "barcode_type",
            "label": _("Barcode Type"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "print_label",
            "label": _("Print Label"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "print_barcode",
            "label": _("Print Barcode Only"),
            "fieldtype": "Data",
            "width": 120
        }
    ]

def get_item_barcode_data(filters):
    """Get items that have barcodes with proper UOM and no duplicates"""
    
    # Build conditions based on filters
    conditions = ["i.disabled = 0", "ib.barcode IS NOT NULL", "ib.barcode != ''"]
    values = []
    
    if filters.get("item_code"):
        conditions.append("i.name = %s")
        values.append(filters.get("item_code"))
    
    if filters.get("item_group"):
        lft, rgt = frappe.db.get_value("Item Group", filters.get("item_group"), ["lft", "rgt"])
        conditions.append("i.item_group IN (SELECT name FROM `tabItem Group` WHERE lft >= %s AND rgt <= %s)")
        values.extend([lft, rgt])
    
    if filters.get("barcode"):
        conditions.append("ib.barcode LIKE %s")
        values.append(f"%{filters.get('barcode')}%")
    
    # Get default price list from settings or use fallback
    default_price_list = get_default_price_list(filters)
    
    # Get company currency
    company = filters.get("company") or frappe.defaults.get_user_default("Company")
    currency = frappe.get_cached_value("Company", company, "default_currency") if company else None
    
    # Main query to get item data with barcodes and selling price
    query = """
        SELECT
            i.name as item_code,
            i.item_name,
            i.item_group,
            COALESCE(ib.uom, i.stock_uom) as uom,
            ib.barcode,
            ib.barcode_type,
            '{currency}' as currency,
            COALESCE(
                (SELECT ip.price_list_rate
                 FROM `tabItem Price` ip
                 WHERE ip.item_code = i.name
                 AND ip.price_list = %s
                 AND ip.selling = 1
                 AND IFNULL(ip.uom, i.stock_uom) = COALESCE(ib.uom, i.stock_uom)
                 AND (ip.valid_from IS NULL OR ip.valid_from <= CURDATE())
                 AND (ip.valid_upto IS NULL OR ip.valid_upto >= CURDATE())
                 ORDER BY ip.valid_from DESC
                 LIMIT 1),
                i.standard_rate,
                0
            ) as selling_price
        FROM
            `tabItem` i
        INNER JOIN
            `tabItem Barcode` ib ON i.name = ib.parent
        WHERE
            {conditions}
        ORDER BY
            i.item_name ASC, ib.idx ASC
    """.format(
        conditions=" AND ".join(conditions),
        currency=currency or frappe.defaults.get_global_default("currency")
    )
    
    # Execute query with price list parameter
    values.insert(0, default_price_list)
    item_data = frappe.db.sql(query, values, as_dict=1)
    
    # Process and clean data
    processed_data = []
    seen_combinations = set()
    
    for row in item_data:
        # Create unique key to prevent duplicates
        key = (row.get("item_code"), row.get("barcode"), row.get("uom"))
        if key in seen_combinations:
            continue
        seen_combinations.add(key)
        
        # Clean None values
        for field in row:
            if row[field] is None:
                if field == "selling_price":
                    row[field] = 0.0
                else:
                    row[field] = ""
        
        # Ensure selling_price is float
        row["selling_price"] = flt(row.get("selling_price", 0))
        
        processed_data.append(row)
    
    return processed_data

def get_default_price_list(filters):
    """Get the default price list based on filters or settings"""
    if filters.get("price_list"):
        return filters.get("price_list")
    
    # Try to get from Selling Settings
    price_list = frappe.db.get_single_value("Selling Settings", "selling_price_list")
    
    # Fallback to Standard Selling
    if not price_list:
        price_list = frappe.db.get_value("Price List", 
            {"selling": 1, "enabled": 1}, 
            "name", 
            order_by="creation"
        ) or "Standard Selling"
    
    return price_list

def get_summary(data):
    """Generate summary statistics for the report"""
    if not data:
        return []
    
    total_items = len(set(d.get("item_code") for d in data))
    total_barcodes = len(data)
    items_with_price = len([d for d in data if flt(d.get("selling_price", 0)) > 0])
    items_without_price = total_barcodes - items_with_price
    
    return [
        {
            "value": total_items,
            "label": _("Total Unique Items"),
            "datatype": "Int",
        },
        {
            "value": total_barcodes,
            "label": _("Total Barcodes"),
            "datatype": "Int",
        },
        {
            "value": items_with_price,
            "label": _("Items with Price"),
            "datatype": "Int",
        },
        {
            "value": items_without_price,
            "label": _("Items without Price"),
            "datatype": "Int",
            "indicator": "red" if items_without_price > 0 else "green"
        }
    ]